garak/_plugins.py

```diff
@@ -5,24 +5,207 @@
 
 import importlib
 import inspect
+import json
 import logging
+import shutil
 import os
-from typing import List
+from typing import List, Callable, Union
+from pathlib import Path
 
 from garak import _config
 from garak.exception import GarakException
 
 PLUGIN_TYPES = ("probes", "detectors", "generators", "harnesses", "buffs")
 PLUGIN_CLASSES = ("Probe", "Detector", "Generator", "Harness", "Buff")
+TIME_FORMAT = "%Y-%m-%d %H:%M:%S %z"
 
 
-@staticmethod
-def _extract_modules_klasses(base_klass):
-    return [  # Extract only classes with same source package name
-        name
-        for name, klass in inspect.getmembers(base_klass, inspect.isclass)
-        if klass.__module__.startswith(base_klass.__name__)
-    ]
+class PluginEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, set):
+            return list(obj)  # allow set as list
+        if isinstance(obj, Path):
+            # relative path for now, may be better to suppress `Path` objects
+            return str(obj).replace(str(_config.transient.basedir), "")
+        try:
+            return json.JSONEncoder.default(self, obj)
+        except TypeError as e:
+            logging.debug("Attempt to serialize JSON skipped: %s", e)
+            return None  # skip items that cannot be serialized at this time
+
+
+class PluginCache:
+    _plugin_cache_file = _config.transient.basedir / "resources" / "plugin_cache.json"
+    _user_plugin_cache_file = _plugin_cache_file
+    _plugin_cache_dict = None
+
+    def __init__(self) -> None:
+        if PluginCache._plugin_cache_dict is None:
+            PluginCache._plugin_cache_dict = self._load_plugin_cache()
+
+    @staticmethod
+    def _extract_modules_klasses(base_klass):
+        return [  # Extract only classes with same source package name
+            name
+            for name, klass in inspect.getmembers(base_klass, inspect.isclass)
+            if klass.__module__.startswith(base_klass.__name__)
+        ]
+
+    def _load_plugin_cache(self):
+        if not os.path.exists(self._plugin_cache_file):
+            self._build_plugin_cache()
```

Comment:
it looks like the cache is only updated if the cache file is absent, so if a file gets modified, the cache data is stale. i remember the intent is to put user plugins elsewhere and assume there's a pip install that isn't touched. on the other hand, this makes life slightly worse for us garak devs - testing whether new plugins load within garak becomes harder, and one can get caught out by forgetting to zero out the cache file every time; the alternative is no cache which also makes our work harder, because we get a delay in the inside loop of our dev cycle. did i get it right? is there something more we can do? i'm noticing those datestamps....

Reply:
You have this right, this first iteration does not complete the cache-mis and developer case, this should not create a significant issue for developer as I don't expect `list` operations to be a common testing practice and any missing cache item will simply pass thru to to import based lookup.

A new plugin will not be found in the cache and a modified plugin in theory already is being loaded by the command used to test any changes. This *could* impact devs to some extent however the limited impact should be reasonable and a follow on improvement should come quickly as the user data location is next on the list to tackle.

Reply:
Yeah, you're right about list operations.

On this topic - what are the steps to rebuild the cache, so I can add them to the pypi build workflow?

Reply:
The CI/CD steps are in the description for this PR, however they should not need to be performed as part of the pypi build process unless you are going to commit the updated file before tagging the release to for build.  The goal will be for automation to keep this file up to date on `main` at all times via triggers on land of each PR.